src/chain/link.rs

```diff
@@ -0,0 +1,238 @@
+        let (access, layout, stages, (src, dst)) = match *self {
+            LinkSync::None(_) | LinkSync::Semaphore { .. } => {
+                return;
+            }
+            LinkSync::Barrier {
+                ref access,
+                ref layout,
+                ref stages,
+            } => (access, layout, stages, (this.family(), this.family())),
+            LinkSync::BarrierSemaphore {
+                ref access,
+                ref layout,
+                ref stages,
+                ..
+            } => (access, layout, stages, (this.family(), this.family())),
+            LinkSync::Transfer {
+                ref access,
+                ref layout,
+                ref stages,
+                other,
+                ..
+            } => (
+                access,
+                layout,
+                stages,
+                M::src_dst(this.family(), other.family()),
+            ),
+        };
+        if src != dst {
+            unimplemented!();
+        }
+        match resources {
+            Some(resources) => {
+                commands.pipeline_barrier(
+                    stages.clone(),
+                    Dependencies::empty(),
+                    resources.iter().map(|&(resource, ref range)| {
+                        resource.barrier(access.clone(), layout.clone(), R::Range::clone(range))
+                    }),
+                );
+            }
+            None => {
+                assert_eq!(
+                    layout.start, layout.end,
+                    "Can't use big barrier if layout transition is required"
+                );
+                assert_eq!(
+                    src, dst,
+                    "Can't use big barrier if ownership transfer is required"
+                );
+                commands.pipeline_barrier(
+                    stages.clone(),
+                    Dependencies::empty(),
+                    Some(R::big_barrier(access.clone())),
+                );
+            }
+        }
+    }
+}
+
+/// Link of the resource chain.
+/// Link corresponds to single rendering-pass or similar entity.
+/// Rendering pass is expected to use resources associated with chain only in ways specified in the link.
+#[derive(Clone, Debug)]
+pub struct Link<A, L, S, W> {
+    pub(super) queue: QueueId,
+    pub(super) stages: PipelineStage,
+    pub(super) access: A,
+    pub(super) layout: L,
+    pub(super) merged_access: A,
+    pub(super) merged_layout: L,
+    pub(super) merged_stages: PipelineStage,
+    pub(super) acquire: LinkSync<A, L, W, Acquire>,
+    pub(super) release: LinkSync<A, L, S, Release>,
+}
+
+impl<A, L, S, W> Link<A, L, S, W>
+where
+    A: Access,
+    L: Layout,
+{
+    /// Get allowed access type for the link.
+    pub fn access(&self) -> A {
+        self.access
+    }
+
+    /// Get resource layout for the link.
+    pub fn layout(&self) -> L {
+        self.layout
+    }
+
+    /// Record acquire barrier if required.
+    pub fn acquire<R, C>(
```

Comment:
similarly, the `acquire` and `release` could be moved out into a separate `impl` block parametrized by `R`